File: Daily Resources/Day11/DSA - Tree Solutions/company_queries_ii.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 5, LOG = 20;
vector<int> tree[N];
int up[N][LOG], depth[N];

void dfs(int u, int p) {
    up[u][0] = p;
    for (int i = 1; i < LOG; ++i) {
        if (up[u][i-1] != -1)
            up[u][i] = up[up[u][i-1]][i-1];
    }
    for (int v : tree[u]) {
        if (v == p) continue;
        depth[v] = depth[u] + 1;
        dfs(v, u);
    }
}

int lca(int a, int b) {
    if (depth[a] < depth[b]) swap(a, b);
    for (int i = LOG - 1; i >= 0; --i)
        if (depth[a] - (1 << i) >= depth[b])
            a = up[a][i];
    if (a == b) return a;
    for (int i = LOG - 1; i >= 0; --i) {
        if (up[a][i] != -1 && up[a][i] != up[b][i]) {
            a = up[a][i];
            b = up[b][i];
        }
    }
    return up[a][0];
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    int n, q;
    cin >> n >> q;
    for (int i = 2, b; i <= n; ++i) {
        cin >> b;
        tree[b].push_back(i);
    }

    dfs(1, -1);

    while (q--) {
        int a, b;
        cin >> a >> b;
        cout << lca(a, b) << '\n';
    }
}

File: Daily Resources/Day11/DSA - Tree Solutions/company_queries_ii.py
import sys
input = sys.stdin.read
sys.setrecursionlimit(2 * 10**5 + 10)

data = input().split()
n, q = int(data[0]), int(data[1])
boss = [0] + [int(x) for x in data[2:n]]
queries = data[n:]

LOG = 20
up = [[-1] * LOG for _ in range(n + 1)]
depth = [0] * (n + 1)
tree = [[] for _ in range(n + 1)]

for i in range(2, n + 1):
    tree[boss[i - 1]].append(i)

def dfs(u, p):
    up[u][0] = p
    for i in range(1, LOG):
        if up[u][i-1] != -1:
            up[u][i] = up[up[u][i-1]][i-1]
    for v in tree[u]:
        depth[v] = depth[u] + 1
        dfs(v, u)

dfs(1, -1)

def lca(a, b):
    if depth[a] < depth[b]:
        a, b = b, a
    for i in range(LOG - 1, -1, -1):
        if depth[a] - (1 << i) >= depth[b]:
            a = up[a][i]
    if a == b:
        return a
    for i in range(LOG - 1, -1, -1):
        if up[a][i] != -1 and up[a][i] != up[b][i]:
            a = up[a][i]
            b = up[b][i]
    return up[a][0]

res = []
for i in range(0, len(queries), 2):
    a = int(queries[i])
    b = int(queries[i+1])
    res.append(str(lca(a, b)))

print('\n'.join(res))
